activation/handler.go

```diff
@@ -139,112 +139,112 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.ActivationTx) error
-			return errors.New("initial Post indices included in challenge does not equal to the initial Post indices included in the atx")
+			return nil, fmt.Errorf("initial Post indices included in challenge does not equal to the initial Post indices included in the atx")
 		}
 
 		// Use the NIPost's Post metadata, while overriding the challenge to a zero challenge,
 		// as expected from the initial Post.
 		initialPostMetadata := *atx.NIPost.PostMetadata
 		initialPostMetadata.Challenge = shared.ZeroChallenge
 		if err := h.nipostValidator.ValidatePost(atx.NodeID().ToBytes(), atx.InitialPost, &initialPostMetadata, uint(atx.NumUnits)); err != nil {
-			return fmt.Errorf("invalid initial Post: %v", err)
+			return nil, fmt.Errorf("invalid initial Post: %v", err)
 		}
 	}
 	var baseTickHeight uint64
 	if atx.PositioningATX != h.goldenATXID {
 		posAtx, err := h.cdb.GetAtxHeader(atx.PositioningATX)
 		if err != nil {
-			return fmt.Errorf("positioning atx not found")
+			return nil, fmt.Errorf("positioning atx not found")
 		}
 		if !atx.PubLayerID.After(posAtx.PubLayerID) {
-			return fmt.Errorf("atx layer (%v) must be after positioning atx layer (%v)",
-				atx.PubLayerID, posAtx.PubLayerID)
+			return nil, fmt.Errorf("atx layer (%v) must be after positioning atx layer (%v)", atx.PubLayerID, posAtx.PubLayerID)
 		}
 		if d := atx.PubLayerID.Difference(posAtx.PubLayerID); d > h.layersPerEpoch {
-			return fmt.Errorf("expected distance of one epoch (%v layers) from pos atx but found %v",
-				h.layersPerEpoch, d)
+			return nil, fmt.Errorf("expected distance of one epoch (%v layers) from pos atx but found %v", h.layersPerEpoch, d)
 		}
 		baseTickHeight = posAtx.TickHeight()
 	} else {
 		publicationEpoch := atx.PubLayerID.GetEpoch()
 		if !publicationEpoch.NeedsGoldenPositioningATX() {
-			return fmt.Errorf("golden atx used for atx in epoch %d, but is only valid in epoch 1", publicationEpoch)
+			return nil, fmt.Errorf("golden atx used for atx in epoch %d, but is only valid in epoch 1", publicationEpoch)
 		}
 	}
 
 	expectedChallengeHash, err := atx.NIPostChallenge.Hash()
 	if err != nil {
-		return fmt.Errorf("failed to compute NIPost's expected challenge hash: %v", err)
+		return nil, fmt.Errorf("failed to compute NIPost's expected challenge hash: %v", err)
 	}
 
 	h.log.WithContext(ctx).With().Info("validating nipost", log.String("expected_challenge_hash", expectedChallengeHash.String()), atx.ID())
 
 	pubKey := signing.NewPublicKey(atx.NodeID().ToBytes())
 	leaves, err := h.nipostValidator.Validate(*pubKey, atx.NIPost, *expectedChallengeHash, uint(atx.NumUnits))
 	if err != nil {
-		return fmt.Errorf("invalid nipost: %v", err)
+		return nil, fmt.Errorf("invalid nipost: %v", err)
 	}
-	atx.Verify(baseTickHeight, leaves/h.tickSize)
-	return nil
+
+	vAtx := atx.Verify(baseTickHeight, leaves/h.tickSize)
+
+	// TODO(mafa): instead of reporting the event immediately, should we only report it after validation passed?
+	events.ReportNewActivation(vAtx)
```

Comment:
i don't know why, but i think the right thing to do is to report it after ProcessAtx() succeed, before "got new atx" logged.